File: leetcode/best_time_to_buy_and_sell_stockIII.js
/* Say you have an array for which the ith element is the price of a given stock
 * on day i.
 * Design an algorithm to find the maximum profit. You may complete at most two 
 * transactions.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    var buy1 = -Math.pow(2, 31), buy2 = -Math.pow(2, 31)
    for (var sell1 = 0, sell2 = 0, i = 0; i < prices.length; ++i) {
        sell2 = Math.max(sell2, buy2 + prices[i])
        buy2 = Math.max(buy2, sell1 - prices[i])
        sell1 = Math.max(sell1, buy1 + prices[i])
        buy1 = Math.max(buy1, -prices[i])
    }
    return sell2
};

File: lintcode/best_time_to_buy_and_sell_stockIII.py
class Solution:
    """
    @param prices: Given an integer array
    @return: Maximum profit
    """
    def maxProfit(self, prices):
        buy1, buy2, sell1, sell2 = -2 ** 32, -2 ** 32, 0, 0
        for price in prices:
            buy2, sell2 = max(buy2, sell1 - price), max(sell2, buy2 + price)
            buy1, sell1 = max(buy1,       - price), max(sell1, buy1 + price)
        return sell2
